packages/shared/src/telemetry/collector.ts

```diff
@@ -0,0 +1,300 @@
+  endpoint: string;
+};
+
+type TelemetryMetadata = Required<
+  Pick<TelemetryCollectorOptions, 'clerkVersion' | 'sdk' | 'sdkVersion' | 'publishableKey' | 'secretKey'>
+> & {
+  /**
+   * The instance type, derived from the provided publishableKey.
+   */
+  instanceType: InstanceType;
+};
+
+type TelemetryEvent = {
+  event: string;
+  /**
+   * publishableKey
+   */
+  pk?: string;
+  /**
+   * secretKey
+   */
+  sk?: string;
+  /**
+   * instanceType
+   */
+  it: InstanceType;
+  /**
+   * clerkVersion
+   */
+  cv: string;
+  /**
+   * SDK
+   */
+  sdk?: string;
+  /**
+   * SDK Version
+   */
+  sdkv?: string;
+  payload: Record<string, string | number | boolean>;
+};
+
+const DEFAULT_CONFIG: Partial<Required<TelemetryCollectorConfig>> = {
+  samplingRate: 1,
+  maxBufferSize: 5,
+  // Production endpoint: https://clerk-telemetry.com
+  // Staging endpoint: https://staging.clerk-telemetry.com
+  // Local: http://localhost:8787
+  endpoint: 'https://clerk-telemetry.com',
+};
+
+export class TelemetryCollector {
+  #config: Required<TelemetryCollectorConfig>;
+  #metadata: TelemetryMetadata = {} as TelemetryMetadata;
+  #buffer: TelemetryEvent[] = [];
+  #pendingFlush: any;
+
+  constructor(options: TelemetryCollectorOptions) {
+    this.#config = {
+      maxBufferSize: options.maxBufferSize ?? DEFAULT_CONFIG.maxBufferSize,
+      samplingRate: options.samplingRate ?? DEFAULT_CONFIG.samplingRate,
+      disabled: options.disabled ?? false,
+      debug: options.debug ?? false,
+      endpoint: DEFAULT_CONFIG.endpoint,
+    } as Required<TelemetryCollectorConfig>;
+
+    if (!options.clerkVersion && typeof window === 'undefined') {
+      // N/A in a server environment
+      this.#metadata.clerkVersion = '';
+    } else {
+      this.#metadata.clerkVersion = options.clerkVersion ?? '';
+    }
+
+    // We will try to grab the SDK data lazily when an event is triggered, so it should always be defined once the event is sent.
+    this.#metadata.sdk = options.sdk!;
+    this.#metadata.sdkVersion = options.sdkVersion!;
+
+    this.#metadata.publishableKey = options.publishableKey ?? '';
+
+    const parsedKey = parsePublishableKey(options.publishableKey);
+    if (parsedKey) {
+      this.#metadata.instanceType = parsedKey.instanceType;
+    }
+
+    if (options.secretKey) {
+      // Only send the first 16 characters of the secret key to to avoid sending the full key. We can still query against the partial key.
+      this.#metadata.secretKey = options.secretKey.substring(0, 16);
+    }
+  }
+
+  get isEnabled(): boolean {
+    if (this.#metadata.instanceType !== 'development') {
+      return false;
+    }
+
+    // In browser or client environments, we most likely pass the disabled option to the collector, but in environments
+    // where environment variables are available we also check for `CLERK_TELEMETRY_DISABLED`.
+    if (this.#config.disabled || (typeof process !== 'undefined' && isTruthy(process.env.CLERK_TELEMETRY_DISABLED))) {
```

Comment:
I'm nitpicking here 🙃  but in general, I tend to avoid double negatives like `CLERK_TELEMETRY_DISABLED=τρθε` as I find them slightly confusing. I think that positive conditions like `CLERK_TELEMETRY_ENABLED=false` are a bit easier to work with. How do you feel about renaming this to `CLERK_TELEMETRY_ENABLED` and having the users disable telemetry by passing `0` or `false`?

Reply:
I think I prefer `CLERK_TELEMETRY_DISABLED` here, as the default is that it is enabled, so when we set it the value would always be false.

FWIW, I modeled it after how Next.js and Astro handle it:
* https://nextjs.org/telemetry#how-do-i-opt-out
* https://astro.build/telemetry/

Reply:
Ah nice, didn't know about the Next/Astro vars. That makes sense! Resolved.